### PR TITLE
Tickets/dm 26754

### PR DESCRIPTION
**Pull request by iagaponenko**

### CODE REVIEW

File: config/PT1.1/Object.json [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+                "yE2_SG_Sigma",
+                "yRadius_SG",
+                "yRadius_SG_Sigma",
+                "yFlags",
+                "chunkId",
+                "subChunkId"
+            ]
+        }
+    }
+}
```
Comment (n8pease) [inline config/PT1.1/Object.json]: no newline at end of file

File: config/PT1.1/RefObjMatch.json [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+                "angSep",
+                "nRefMatches",
+                "nObjMatches",
+                "closestToRef",
+                "closestToObj",
+                "flags"
+            ]
+        }
+    }
+}
```
Comment (n8pease) [inline config/PT1.1/RefObjMatch.json]: no newline

File: config/PT1.1/SimRefObject.json [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+                "uCov",
+                "gCov",
+                "rCov",
+                "iCov",
+                "zCov",
+                "yCov"
+            ]
+        }
+    }
+}
```
Comment (n8pease) [inline config/PT1.1/SimRefObject.json]: it looks like all these json files are missing newlines. I'll stop mentioning it.

File: include/lsst/partition/ConfigStore.h [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+ * Parameters are fetched from the store using their "path" specification. The dot
+ * character '.' is used for separating components of the path:
+ * @code
+ * foo.bar.p1
+ * dir.p2
+ * p3
+ * @code
+ * The intermediate elements of the path would directly map to the corresponding keys
+ * of the nested JSON objects in the store. For example, the above shown paths would
+ * be represented by the following JSON structure passed into the class's constructr,
```
Comment (n8pease) [inline include/lsst/partition/ConfigStore.h]: constructr -> constructor

File: include/lsst/partition/ConfigStore.h [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+ *    },
+ *    "p3": <value>
+ * }
+ * @code
+ * Values of the command-line parameters are expected to be injected into the configuration
+ * by calling method set(). The long names (excluding preceeding "--") of the parameters
+ * would be treated as above explained path names.
+ *
+ * @note the class's implementation is not thread-safe, and any subsequent modification
+ *   to a parameter would always override its previous value (if any existed).
```
Comment (n8pease) [inline include/lsst/partition/ConfigStore.h]: I think these are unrelated points? 
1. It is not thread-safe (unprotected access could crash)
2. Modifications are last-one wins. 
If so, it probably deserves separate sentences for each point, for clarity.

File: include/lsst/partition/ConfigStore.h [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+     * the object.
+     * 
+     * It's allowed to call this method many times for the same or different files.
+     * Parameters read from a file will be merged into the store.
+     *
+     * @param filename  The name of a JSON file to be parsed.
+     * @throw std::invalid_argument  The file name is not valid.
+     * @throw std::runtime_error  Failed to open or parse the file.
+     */
+    void parse(std::string const& filename);
```
Comment (n8pease) [inline include/lsst/partition/ConfigStore.h:188]: I'd consider naming this as an overload of the `add` function, since it parses and adds. (the other two add functions also do a read-and-add type of behavior).

File: include/lsst/partition/ConfigStore.h [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+    /**
+     * This method should be used instead of 'has(path)' or 'get<bool>(path)' for checking
+     * a status of parameters meant to be used as flags. A reason why this method is needed is
+     * because JSON allows attributes of the 'bool' type which could be explicitly set to 'true'
+     * or 'false'. In this case just relying on 'has(path)' may produce incorrect results.
+     * Method 'has(path)' would only work for parameters found in the command line.
+     * On the other hand, relying on 'get<bool>(path)' would only work for the parameters
+     * which were explicitly set from JSON. Using the getter method would result in an exception
+     * if the parameter wasn't explicitly specified in the JSON file.
+     *
```
Comment (n8pease) [inline include/lsst/partition/ConfigStore.h:254]: Can you come up with a more elegant solution where the caller doesn't have to pick the correct function based on expected type? As is, this is error prone.

File: include/lsst/partition/MapReduce.h [hunk 8]
```diff
@@ -536,7 +537,7 @@ namespace detail {
         std::vector<SortedRecordRange> ranges;
         ranges.reserve(_numWorkers);
         // BEWARE: a lock on _mutex is held while constructing Worker.
-        // This is because _vm contains strings and the current GCC libstc++
+        // This is because _config contains strings and the current GCC libstc++
```
Comment (n8pease) [inline include/lsst/partition/MapReduce.h:540]: Is this whole comment still relevant? Maybe it's enough to say "because ConfigStore is not thread safe"?

File: src/Chunker.cc [hunk 9]
```diff
@@ -69,11 +70,11 @@ Chunker::Chunker(double overlap,
     _initialize(overlap, numStripes, numSubStripesPerStripe);
 }
 
-Chunker::Chunker(po::variables_map const & vm) {
+Chunker::Chunker(ConfigStore const & config) {
     // Overlap is forced to be disabled if using the class in non-RA/DEC paritioning mode
-    _initialize(vm.count("part.pos") == 0 ? 0. : vm["part.overlap"].as<double>(),
-                vm["part.num-stripes"].as<int32_t>(),
-                vm["part.num-sub-stripes"].as<int32_t>());
+    _initialize(!config.has("part.pos") ? 0. : config.get<double>("part.overlap"),
```
Comment (n8pease) [inline src/Chunker.cc]: for readability, why not drop the `!` and reverse the result values?

File: include/lsst/partition/ConfigStore.h (hunk 7 above)
Comment (iagaponenko) [reply include/lsst/partition/ConfigStore.h:254]: I hear two questions here:

1) If it would be better to rely on some form of the method overloading based on a return type.

I know some libraries do that, including `nlohmann::json`. Personally, I don't find this safe as this solution has many corner cases and unpleasant surprises. To me, passing a desired type explicitly to a parameter's value retreival method `T get<T>(name)` seems like the best solution.

Besides, when I was migratng the package to the new `ConfigStore` API I was trying to preserve the original code which uses type specifications.

2) If having method `bool flag(name)` is the right approach compared to some "more elegant solution".

The answer is found in the comment to the method. The original code was relying on "counting" of the number of command line parameters of the given name. This doesn't work with `JSON` where a boolean parameter a) may be present or absent, or b) it may be set to `true` or `false`. So, I decided the safest approach would be to express user's intent as a request for a presence of some "flag" which could be `true` or `false`, regardless if it was found in the command line, or if it was (or was ot) found in ` `JSON` file. To me a concept of the "flag" is close to the actual intent of the client's code than someting like `bool get<bool>(name)`.

File: include/lsst/partition/ConfigStore.h (hunk 6 above)
Comment (iagaponenko) [reply include/lsst/partition/ConfigStore.h:188]: In general, this is good suggestion. Unfortunately, it wouldn't work here since `nlohman::json` seems to have a non-explicit constructor from a string. This cases the following compilation problem:
```
tests/configStore.cc: In member function 'void ConfigStoreTest::test_method()':
tests/configStore.cc:325:29: error: call of overloaded 'add(const char [1])' is ambiguous
         anotherConfig.add("");
                             ^
In file included from tests/configStore.cc:25:0:
include/lsst/partition/ConfigStore.h:188:10: note: candidate: void lsst::partition::ConfigStore::add(const string&)
     void add(std::string const& filename);
          ^~~
include/lsst/partition/ConfigStore.h:200:10: note: candidate: void lsst::partition::ConfigStore::add(const json&)
     void add(nlohmann::json const& config);
          ^~~
```
So, I'm reversing it back to `::parse`.

File: include/lsst/partition/MapReduce.h (hunk 8 above)
Comment (iagaponenko) [reply include/lsst/partition/MapReduce.h:540]: I don't fully understand yet the `MapReduce` implementation to remove the lock on the mutex. The new config system certainly has strings in its implementation. Let's just keep the lock for now until the next major refactoring of the package.
